File: src/main/java/RoleListener.java
import net.dv8tion.jda.core.entities.Guild;
import net.dv8tion.jda.core.entities.Member;
import net.dv8tion.jda.core.entities.Role;
import net.dv8tion.jda.core.entities.User;
import net.dv8tion.jda.core.events.message.guild.GuildMessageReceivedEvent;
import net.dv8tion.jda.core.hooks.ListenerAdapter;
import net.dv8tion.jda.core.managers.GuildController;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

public class RoleListener extends ListenerAdapter {
  Logger logger = LoggerFactory.getLogger(RoleListener.class);

  @Override
  public void onGuildMessageReceived(GuildMessageReceivedEvent event) {
    if (event.getAuthor().isBot()) return;

    User author = event.getAuthor();
    Guild guild = event.getGuild();
    GuildController mod = new GuildController(event.getGuild());
    String roleName = "cool role"; // TODO: make configurable with a command
    List<Role> roles = guild.getRolesByName(roleName, true);

    logger.debug("{} roles found", roles.size());

    if (roles.size() == 0) {
      logger.error("Couldn't find role {}", roleName);
      return;
    }

    Member member = guild.getMember(author);

    logger.trace("member.getEffectiveName() = " + member.getEffectiveName());

    mod.addSingleRoleToMember(member, roles.get(0)).queue();
  }

}

File: src/main/java/com/dongtronic/diabot/listener/RoleListener.kt
package com.dongtronic.diabot.listener

import com.dongtronic.diabot.data.RewardDAO
import com.dongtronic.diabot.util.RoleUtils
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import org.slf4j.LoggerFactory

class RoleListener : ListenerAdapter() {
    private val logger = LoggerFactory.getLogger(RoleListener::class.java)

    override fun onGuildMessageReceived(event: GuildMessageReceivedEvent) {
        if (event.author.isBot) return

        val author = event.author
        val guild = event.guild
        val member = guild.getMember(author) ?: return

        val userRoles = member.roles

        if(RewardDAO.getInstance().getOptOut(guild.id, author.id)) {
            // Skip users that have opted out
            return
        }

        val potentialRewards = RewardDAO.getInstance().getSimpleRewards(guild.id)

        val rewards = RoleUtils.buildRewardsMap(potentialRewards, guild)

        // Check if user applies for new rewards
        for ((required, rewardList) in rewards) {
            if (userRoles.contains(required)) {
                for(reward in rewardList) {
                    if(!userRoles.contains(reward)) {
                        guild.addRoleToMember(member, reward).queue()
                        logger.info("Assigning reward roles ${reward.name} (${reward.id}) to ${author.name}")
                    }
                }
            }
        }
    }


}
